doc-content/jbpm-docs/src/main/asciidoc/Persistence/variables-auditing-con.adoc

```diff
@@ -0,0 +1,156 @@
+
+    /**
+     * Tests if a given variable can be indexed by this indexer
+     * 
+     * NOTE: only one indexer can be used for a given variable
+     * 
+     * @param variable variable to be indexed
+     * @return true if the variable should be indexed with this indexer
+     */
+    boolean accept(Object variable);
+    
+    /**
+     * Performs an index/transform operation of the variable. The result of this operation can be
+     * either a single value or a list of values, to support complex type separation.
+     * For example, when the variable is of the type Person that has name, address, and phone fields,
+     * the indexer could build three entries out of it to represent individual fields:
+     * person = person.name
+     * address = person.address.street
+     * phone = person.phone
+     * this configureation allows advanced queries for finding relevant entries.
+     * @param name name of the variable
+     * @param variable actual variable value 
+     * @return
+     */
+    List<V> index(String name, Object variable);
+}
+----
+
+The default indexer uses the `toString()` method to produce a single audit entry for a single variable. Other indexers can returns a list of objects from indexing a single variable.
+
+To enable efficient queries for the `Person` type, you can build a custom indexer that indexes a `Person` instance into separate audit entries, one representing name and another representing age. 
+
+.Sample indexer for the `Person` type
+[source,java]
+----
+public class PersonTaskVariablesIndexer implements TaskVariableIndexer {
+
+    @Override
+    public boolean accept(Object variable) {
+        if (variable instanceof Person) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public List<TaskVariable> index(String name, Object variable) {
+        
+        Person person = (Person) variable;
+        List<TaskVariable> indexed = new ArrayList<TaskVariable>();
+        
+        TaskVariableImpl personNameVar = new TaskVariableImpl();
+        personNameVar.setName("person.name");
+        personNameVar.setValue(person.getName());
+        
+        indexed.add(personNameVar);
+        
+        TaskVariableImpl personAgeVar = new TaskVariableImpl();
+        personAgeVar.setName("person.age");
+        personAgeVar.setValue(person.getAge()+"");
+        
+        indexed.add(personAgeVar);
+        
+        return indexed;
+    }
+
+}
+----
+
+The {PROCESS_ENGINE} can use this indexer to index values when they are of the `Person` type, while all other variables are indexed with the default `toString()` method. Now, to query for process instances or tasks that refer to a person with age 34, you can use the following query:
+
+* variable name: person.age
+* variable value: 34
+
+As a `LIKE` type query is not used, the database server can optimize the query and make it efficient on a big set of data. 
+
+[float]
+== Building and registering custom indexers
```

Comment:
I would actually convert this to a legitimate 3-step proc (or so) as a dedicated module, and retitle it slightly, as follows

= Building and registering custom indexers for variables

With the proc being:

1. Implement the interfaces to build the custom indexers
2. Implement the two methods
3. Package the implementation

Reason is that, although we can concept-ize very basic procs that would otherwise be clunky as proc modules (like the section before this), I'd say that this one is a stretch and actually yields itself well to be a legit proc.

Reply:
Following a discussion on chat, just changed the title to "Custom indexers", because the steps are too vague for a proper procedure in my view.